File: tests/macos/InteractionBridge.m
#import "InteractionBridge.h"
#import <objc/runtime.h>
#include <string.h>

static bool confirm_enabled = true;
static bool accept_close;
static size_t alerts;
static bool safe_close_default;
static const char *active_theme = "Dark";
static NSString *config_path;
static bool fullscreen;

static NSModalResponse respondToAlert(id alert, SEL selector) {
    alerts++;
    NSAlert *confirmation = alert;
    [confirmation layout];
    safe_close_default = confirmation.buttons.count == 2 &&
        [confirmation.buttons[0].title isEqualToString:@"Cancel"] &&
        [confirmation.buttons[1].title isEqualToString:@"Close"] &&
        confirmation.window.defaultButtonCell != confirmation.buttons[1].cell &&
        ![confirmation.buttons[1].keyEquivalent isEqualToString:@"\r"];
    return accept_close ? NSAlertSecondButtonReturn : NSAlertFirstButtonReturn;
}

void interaction_test_confirmation(bool enabled, bool accept) {
    confirm_enabled = enabled;
    accept_close = accept;
    alerts = 0;
    safe_close_default = false;
    method_setImplementation(class_getInstanceMethod(NSAlert.class, @selector(runModal)), (IMP)respondToAlert);
}
size_t interaction_test_alerts(void) { return alerts; }
bool interaction_test_safe_close_default(void) { return safe_close_default; }
bool mostty_config_confirm_close(void) { return confirm_enabled; }
bool mostty_config_reload(void) { return true; }
void interaction_test_config_path(const char *path) { config_path = [NSString stringWithUTF8String:path]; }
size_t mostty_config_path(uint8_t *buf, size_t cap) {
    const char *path = config_path.UTF8String;
    if (!path || strlen(path) > cap) return 0;
    memcpy(buf, path, strlen(path));
    return strlen(path);
}
bool mostty_config_background_blur(void) { return false; }
bool mostty_config_maximize(void) { return false; }
void interaction_test_fullscreen(bool enabled) { fullscreen = enabled; }
bool mostty_config_fullscreen(void) { return fullscreen; }
size_t mostty_config_launcher_count(void) { return 2; }
static size_t text(const char *value, uint8_t *buf, size_t cap) {
    size_t count = strlen(value);
    if (!cap) return count;
    if (count > cap) return 0;
    memcpy(buf, value, count);
    return count;
}
size_t mostty_config_launcher_text(size_t index, uint32_t field, uint8_t *buf, size_t cap) {
    const char *values[2][3] = {{"First", "echo first", "/usr"}, {"Second", "echo second", "/var"}};
    return index < 2 && field < 3 ? text(values[index][field], buf, cap) : 0;
}
size_t mostty_config_refresh_themes(void) { return 2; }
size_t mostty_config_theme_name(size_t index, uint8_t *buf, size_t cap) {
    return index < 2 ? text(index ? "Light" : "Dark", buf, cap) : 0;
}
size_t mostty_config_active_theme(uint8_t *buf, size_t cap) { return text(active_theme, buf, cap); }
bool mostty_config_select_theme(const char *name) {
    if (strcmp(name, "Dark") == 0) active_theme = "Dark";
    else if (strcmp(name, "Light") == 0) active_theme = "Light";
    else return false;
    return true;
}

File: tests/macos/InteractionBridge.swift
import AppKit
import ObjectiveC

private var confirmEnabled = true
private var acceptClose = false
private var alerts = 0
private var safeCloseDefault = false
private var activeTheme = "Dark"
private var configPath: String?
private var fullscreen = false

private let respondToAlert: @convention(c) (NSAlert, Selector) -> Int = { confirmation, _ in
    alerts += 1
    confirmation.layout()
    safeCloseDefault = confirmation.buttons.count == 2 &&
        confirmation.buttons[0].title == "Cancel" &&
        confirmation.buttons[1].title == "Close" &&
        confirmation.window.defaultButtonCell !== confirmation.buttons[1].cell &&
        confirmation.buttons[1].keyEquivalent != "\r"
    return (acceptClose ? NSApplication.ModalResponse.alertSecondButtonReturn : .alertFirstButtonReturn).rawValue
}
func interaction_test_confirmation(_ enabled: Bool, _ accept: Bool) {
    confirmEnabled = enabled
    acceptClose = accept
    alerts = 0
    safeCloseDefault = false
    let method = class_getInstanceMethod(NSAlert.self, #selector(NSAlert.runModal))!
    method_setImplementation(method, unsafeBitCast(respondToAlert, to: IMP.self))
}
func interaction_test_alerts() -> Int { alerts }
func interaction_test_safe_close_default() -> Bool { safeCloseDefault }
@_cdecl("mostty_config_confirm_close")
func stubConfirmClose() -> Bool { confirmEnabled }
@_cdecl("mostty_config_reload")
func stubConfigReload() -> Bool { true }
func interaction_test_config_path(_ path: String) { configPath = path }
@_cdecl("mostty_config_path")
func stubConfigPath(_ buffer: UnsafeMutablePointer<UInt8>?, _ capacity: Int) -> Int {
    guard let path = configPath else { return 0 }
    let bytes = Array(path.utf8)
    guard bytes.count <= capacity else { return 0 }
    buffer?.update(from: bytes, count: bytes.count)
    return bytes.count
}
@_cdecl("mostty_config_background_blur")
func stubBackgroundBlur() -> Bool { false }
@_cdecl("mostty_config_maximize")
func stubMaximize() -> Bool { false }
func interaction_test_fullscreen(_ enabled: Bool) { fullscreen = enabled }
@_cdecl("mostty_config_fullscreen")
func stubFullscreen() -> Bool { fullscreen }
@_cdecl("mostty_config_launcher_count")
func stubLauncherCount() -> Int { 2 }
private func text(_ value: String, _ buffer: UnsafeMutablePointer<UInt8>?, _ capacity: Int) -> Int {
    let bytes = Array(value.utf8)
    if capacity == 0 { return bytes.count }
    guard bytes.count <= capacity else { return 0 }
    buffer?.update(from: bytes, count: bytes.count)
    return bytes.count
}
@_cdecl("mostty_config_launcher_text")
func stubLauncherText(_ index: Int, _ field: UInt32, _ buffer: UnsafeMutablePointer<UInt8>?, _ capacity: Int) -> Int {
    let values = [["First", "echo first", "/usr"], ["Second", "echo second", "/var"]]
    return index >= 0 && index < 2 && field < 3 ? text(values[index][Int(field)], buffer, capacity) : 0
}
@_cdecl("mostty_config_refresh_themes")
func stubRefreshThemes() -> Int { 2 }
@_cdecl("mostty_config_theme_name")
func stubThemeName(_ index: Int, _ buffer: UnsafeMutablePointer<UInt8>?, _ capacity: Int) -> Int {
    index >= 0 && index < 2 ? text(index == 0 ? "Dark" : "Light", buffer, capacity) : 0
}
@_cdecl("mostty_config_active_theme")
func stubActiveTheme(_ buffer: UnsafeMutablePointer<UInt8>?, _ capacity: Int) -> Int { text(activeTheme, buffer, capacity) }
@_cdecl("mostty_config_select_theme")
func stubSelectTheme(_ name: UnsafePointer<CChar>?) -> Bool {
    guard let name else { return false }
    let theme = String(cString: name)
    guard theme == "Dark" || theme == "Light" else { return false }
    activeTheme = theme
    return true
}
